### PR TITLE
[feat](nereids) add merge aggregate rule

### PR DESCRIPTION
**Pull request by feiniaofeiafei**
```sql
select 
  deptno c, 
  min(y), 
  max(z) z, 
  sum(r), 
  sum(m) n, 
  sum(x) sal 
from 
  (
    select 
      deptno, 
      ename, 
      sum(sal) x, 
      max(sal) z, 
      min(sal) y, 
      count(hiredate) m, 
      count(mgr) r 
    from 
      sales.emp 
    group by 
      deptno, 
      ename
  ) t 
group by 
  deptno;
```
aggregate can be merged to
```sql
select 
  deptno c, 
  min(sal), 
  max(sal) z, 
  count(mgr), 
  count(hiredate) n, 
  sum(sal) sal 
from 
     sales.emp 
group by 
  deptno;
```
this pr add a RBO rule to perform this transform

### CODE REVIEW

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MergeAggregate.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            // not support outerAggFunc: sum(a+1),sum(a+b)
+            if (!(outerFunc.child(0) instanceof SlotReference)) {
+                return false;
+            }
+            ExprId childExprId = ((SlotReference) outerFunc.child(0)).getExprId();
+            if (innerAggExprIdToAggFunc.containsKey(childExprId)) {
+                AggregateFunction innerFunc = innerAggExprIdToAggFunc.get(childExprId);
+                if (innerFunc.isDistinct()) {
+                    return false;
+                }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MergeAggregate.java:160]: inner distinct is ok if outer group by keys are exactly same with inner keys?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MergeAggregate.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                // support sum(sum),min(min),max(max),sum(count),any_value(any_value)
+                if (!(outerFunc.getName().equals("sum") && innerFunc.getName().equals("count"))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MergeAggregate.java:162]: trans `sum(count())` to `count()` lead to nullable changed if outer agg is scalar agg. so we need to wrap the final expression with `nullable()` function to change its nullable to true

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MergeAggregate.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                        && !innerFunc.getName().equals(outerFunc.getName())) {
+                    return false;
+                }
+            } else {
+                if (!outerFunc.getName().equals("max")
+                        && !outerFunc.getName().equals("min")
+                        && !outerFunc.getName().equals("any_value")) {
+                    return false;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MergeAggregate.java:170]: could u add some comment to explain this else branch?

File: regression-test/suites/nereids_rules_p0/merge_aggregate/merge_aggregate.groovy [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+         properties('replication_num' = '1'); 
+         """
+
+    sql """
+         insert into mal_test1 values(2,1,3),(1,1,2),(3,5,6),(6,null,6),(4,5,6);
+      """
+
+    sql "sync"
+
+    qt_maxMax_minMin_sumSum_sumCount """
```
Comment (morrySnow) [inline regression-test/suites/nereids_rules_p0/merge_aggregate/merge_aggregate.groovy:35]: add some shape check or ut to ensure merge agg work well

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MergeAggregate.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        LogicalAggregate<Plan> innerAgg = project.child();
+
+        List<Expression> outerAggGroupByKeys = PlanUtils.replaceExpressionByProjections(project.getProjects(),
+                outerAgg.getGroupByExpressions());
+        if (!new HashSet<>(innerAgg.getGroupByExpressions()).containsAll(outerAggGroupByKeys)) {
+            return false;
+        }
+        // project cannot have expressions like a+1
+        if (ExpressionUtils.anyMatch(project.getProjects(),
+                expr -> !(expr instanceof SlotReference) && !(expr instanceof Alias))) {
```
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MergeAggregate.java:205]: the code allows Alias( a + 1), this is different from comment. Please double check the expected behavior

**Review by starocean999 — APPROVED** [review]

**Pull request merged**